ADR: JavaScript Modules exports convention (#13)
straw-hat-team/adr: adrs/4937890790/README.md

# JavaScript Modules exports convention

* **State:** Draft
* **Created:** 2021-02-17
* **Tags:** js,ts

## Context

We don't have a consistent way to export the modules' functions and variables
across the JavaScript ecosystem.

Although such concern is not a big concern for more Senior developers, there is
some cognitive load when it comes to default exports over named exports.

What is one of the most challenging things in computer science: naming.

Using default-exports over named-exports forces the user to think about the name

> of the import.

Author: > Although such concern is not a big concern for more Senior developers

😄 you are too ahead of most people, I am taking into consideration Bootcamp students, and self-taught people, and so on

Let's use the following example:

```ts
import [name here] from 'base-button'
```

What should be the right name for the import?

* `Button`
* `button`

> * `baseButton`

Author: @TMiller00 good observation, that is actually another reason why it is important to give named-exports. But let me remove that one.

# JavaScript Modules exports convention

* **State:** Draft
* **Created:** 2021-02-17
* **Tags:** js,ts

## Context

> We don't have a consistent way to export a module's functions and variables
> across the JavaScript ecosystem.
> 
> Although such concern is not a big concern for more Senior developers, there is
> some cognitive load when it comes to default exports over named exports.

Reviewer: ```suggestion
We don't have a consistent way to export a module's functions and variables
across the JavaScript ecosystem.  In the modern era of ES6 exports, there are
generally 2 camps:

1. Use `export default` for your "main" export. 
  This option forces a second choice for auxiliary exports, where you can either 
  use named exports or `Object.assign()` the default.
2. Use named exports for everything.

Teams that don't choose a camp are inflicting an unnecessary cognitive load 
on themselves.
```

Wanted to take a stab at laying the options.

What is one of the most challenging things in computer science? Naming.

Using default-exports over named-exports forces the user to think about the name
of the import.

Let's use the following example:

```ts
import [name here] from 'base-button'
```

What should be the right name for the import?

* `Button`
* `baseButton`
* `BaseButton`
* `MyButton` or any other name you would like it to be

> For a simple import, you have already at least three different variations to
> think about it, or combinations of any name for that matter.

Reviewer: ```suggestion
For a simple default import, you already have at least three different variations to
think about and combinations of any name for that matter.
```

> 

Reviewer: I know this comes up later, but I think it's horribly BM to rename imports -- even  in the `export default` world.  😂 

```ts
import Button from 'base-button'
```

In some cases, you have to avoid name conflict.

```ts
import TheirBaseButton from 'base-button'
import OurBaseButton from '../base-button'
```

Besides that, why would anybody do that?

Besides that, in some cases, you even have to think about the context of the
file.

For example, if the `base-button` is in React components, probably the right
import name would be either `Button` or `BaseButton` or anything that it uses
`PascalCase.` That is a limitation of React itself. You can not be "100% sure"
unless you inspect the file and understand the context of such default-export.

It is not a simple task; therefore, people rely on linters and code reviewers to
improve the source code's quality.

> If you already thought about a name and knew the context, and gave it a name
> inside your module, don't force users to think about a new name every time
> they must import the component, most of the case the named you came up with is
> good enough; so you can take advantage of named-exports.

Reviewer: ```suggestion
Naming is hard.  When you are authoring code, do not make the consumers of your
code -- the importers of your exports --  figure out what names to use.  As an author, 
you have the advantage of having thought about its intent and purpose; you are
aware of its context and limitations and most likely gave it a name inside your module.
That name is good enough, and by using named exports, you  are doing your consumers a favor, 
by freeing them from having to rediscover that name. 
```

export function SomethingElse() {
  return (
    <>
      ...
      <Btn />
    </>
  );
}
```

How confident could you be that those two imports are part of the same context
without the cognitive load of reading the imports to create such mapping in your
brain?

Answer: probably close to none.

* You may argue, How is this any different from named-exports?

Take the following files:

File #1:

```ts
import { Button } from "...";

export function Something() {
  return (
    <>
      ...
      <Button />
    </>
  );
}
```

File #2:

```ts
import { Button as Btn } from "...";

export function SomethingElse() {
  return (
    <>
      ...
      <Btn />
    </>
  );
}
```

How confident could you be that those two imports are part of the same context?

Answer: somewhat confident.

Because in the worst case, you can see the mapping from `Button` to `Btn`
happening in the code, although you are not 100% confident that `Btn` is the same
`Button` context, because you still need to know the import, you are somehow
optimistic that it may be.
There is an explicit remapping happening, helping the reader to understand that
in this file, for whatever reason, `Button` must become `Btn`.

> Also, having explicit names allows IDEs to figure out what you are trying to
> accomplish and give you much better autocomplete, refactoring, and automation,
> since those named-exports are static and don't require runtime to figure out
> the names, they are able to index the source code much easier and algorithms
> could take advantage of the explicitness and static nature of such habit.

Reviewer: ```suggestion
Also, having explicit names allows IDEs to figure out what you are trying to
accomplish and give you much better autocomplete refactoring and automation.
Since those named-exports are static and don't require a runtime to figure out
the names, they can index the source code and algorithms
can take advantage of the explicitness and static nature of the names.
```

## Resolution

* You **must** use named-exports over default-exports unless some limitation
  is imposed
* You **must** provide named-export and default-exports for React lazy

>   loadable components.

Reviewer: What do you think about adding a resolution, something like:

You should not rename code unless you're solving a scoping conflict.

Author: I am not following, what do you mean by scoping conflict? and where?
